storage/inmem/txn.go

```diff
@@ -0,0 +1,410 @@
+// Copyright 2017 The OPA Authors.  All rights reserved.
+// Use of this source code is governed by an Apache2
+// license that can be found in the LICENSE file.
+
+package inmem
+
+import (
+	"encoding/json"
+	"fmt"
+	"strconv"
+
+	"github.com/open-policy-agent/opa/storage"
+)
+
+type transaction struct {
+	xid      uint64
+	write    bool
+	log      *snapshot
+	db       *store
+	policies map[string][]byte
+}
+
+type snapshot struct {
+	path   storage.Path
+	remove bool
+	stale  bool
+	value  interface{}
+	next   *snapshot
+	prev   *snapshot
+}
+
+func (txn *transaction) ID() uint64 {
+	return txn.xid
+}
+
+func (txn *transaction) assert(op storage.PatchOp, path storage.Path, value interface{}) error {
+
+	if !txn.write {
+		return &storage.Error{
+			Code:    storage.InvalidTransactionErr,
+			Message: "data write during read transaction",
+		}
+	}
+
+	if len(path) == 0 {
+		return txn.assertRoot(op, value)
+	}
+
+	for curr := txn.log; curr != nil; curr = curr.next {
+		if curr.path.Equal(path) {
+			if curr.remove {
+				if op != storage.AddOp {
+					return notFoundError(path)
+				}
+			}
+			curr.stale = true
+			break
+		} else if curr.path.HasPrefix(path) {
+			curr.stale = true
```

Comment:
This stale case definitely needs a comment: a top level comment around "stale bool" would have been enough.